Make AUTH optional for KafkaSource and KafkaTarget

Changes:
* `auth` should be documented as optional (is documented as required)
* `security mechanisms` is required, can we make it optional
* make sure our default settings match those of a clean install of Kafka or RedPanda

you have to look in the kafka target logs to figure out what actually happened:

```
2022/12/21 17:15:18 Error processing env var: required key SECURITY_MECHANISMS missing value

panic: Error processing env var: required key SECURITY_MECHANISMS missing value
```

--- a/pkg/apis/targets/v1alpha1/deepcopy_generated.go
+++ b/pkg/apis/targets/v1alpha1/deepcopy_generated.go
@@ -2651,7 +2651,11 @@ func (in *KafkaTargetSpec) DeepCopyInto(out *KafkaTargetSpec) {
 		*out = make([]string, len(*in))
 		copy(*out, *in)
 	}
-	in.Auth.DeepCopyInto(&out.Auth)
+	if in.Auth != nil {
+		in, out := &in.Auth, &out.Auth
+		*out = new(KafkaTargetAuth)
+		(*in).DeepCopyInto(*out)
+	}
 	if in.AdapterOverrides != nil {
 		in, out := &in.AdapterOverrides, &out.AdapterOverrides
 		*out = new(commonv1alpha1.AdapterOverrides)

--- a/pkg/apis/targets/v1alpha1/kafka_types.go
+++ b/pkg/apis/targets/v1alpha1/kafka_types.go
@@ -58,7 +58,8 @@ type KafkaTargetSpec struct {
 	BootstrapServers []string `json:"bootstrapServers"`
 
 	// Auth contains Authentication method used to interact with Kafka.
-	Auth KafkaTargetAuth `json:"auth"`
+	// +optional
+	Auth *KafkaTargetAuth `json:"auth"`
 
 	// Whether to omit CloudEvent context attributes in messages sent to Kafka.
 	// When this property is false (default), the entire CloudEvent payload is included.

--- a/pkg/sources/adapter/kafkasource/env.go
+++ b/pkg/sources/adapter/kafkasource/env.go
@@ -28,7 +28,7 @@ func NewEnvConfig() adapter.EnvConfigAccessor {
 type envAccessor struct {
 	adapter.EnvConfig
 
-	SASLEnable bool `envconfig:"SASL_ENABLE" required:"true"`
+	SASLEnable bool `envconfig:"SASL_ENABLE" required:"false"`
 	TLSEnable  bool `envconfig:"TLS_ENABLE" required:"false"`
 
 	BootstrapServers []string `envconfig:"BOOTSTRAP_SERVERS" required:"true"`
@@ -37,7 +37,7 @@ type envAccessor struct {
 	Topic            string   `envconfig:"TOPIC" required:"true"`
 	GroupID          string   `envconfig:"GROUP_ID" required:"false"`
 
-	SecurityMechanisms  string `envconfig:"SECURITY_MECHANISMS" required:"true"`
+	SecurityMechanisms  string `envconfig:"SECURITY_MECHANISMS" required:"false"`
 	KerberosConfigPath  string `envconfig:"KERBEROS_CONFIG_PATH" required:"false" `
 	KerberosServiceName string `envconfig:"KERBEROS_SERVICE_NAME" required:"false" `
 	KerberosKeytabPath  string `envconfig:"KERBEROS_KEYTAB_PATH" required:"false"`

--- a/pkg/targets/adapter/kafkatarget/config.go
+++ b/pkg/targets/adapter/kafkatarget/config.go
@@ -28,7 +28,7 @@ func EnvAccessorCtor() pkgadapter.EnvConfigAccessor {
 type envAccessor struct {
 	pkgadapter.EnvConfig
 
-	SASLEnable bool `envconfig:"SASL_ENABLE" required:"true"`
+	SASLEnable bool `envconfig:"SASL_ENABLE" required:"false"`
 	TLSEnable  bool `envconfig:"TLS_ENABLE" required:"false"`
 
 	BootstrapServers   []string `envconfig:"BOOTSTRAP_SERVERS" required:"true"`
@@ -36,7 +36,7 @@ type envAccessor struct {
 	Password           string   `envconfig:"PASSWORD" required:"false"`
 	Topic              string   `envconfig:"TOPIC" required:"true"`
 	GroupID            string   `envconfig:"GROUP_ID" required:"false" `
-	SecurityMechanisms string   `envconfig:"SECURITY_MECHANISMS" required:"true"`
+	SecurityMechanisms string   `envconfig:"SECURITY_MECHANISMS" required:"false"`
 
 	KerberosConfigPath  string `envconfig:"KERBEROS_CONFIG_PATH" required:"false" `
 	KerberosServiceName string `envconfig:"KERBEROS_SERVICE_NAME" required:"false" `

--- a/pkg/targets/reconciler/kafkatarget/adapter.go
+++ b/pkg/targets/reconciler/kafkatarget/adapter.go
@@ -77,27 +77,29 @@ func (r *Reconciler) BuildAdapter(trg commonv1alpha1.Reconcilable, _ *apis.URL) 
 	var secretVolumes []corev1.Volume
 	var secretVolMounts []corev1.VolumeMount
 
-	if typedTrg.Spec.Auth.Kerberos != nil {
-		if typedTrg.Spec.Auth.Kerberos.Config != nil {
-			configVol, configVolMount := secretVolumeAndMountAtPath(
-				"krb5-config",
-				krb5ConfPath,
-				typedTrg.Spec.Auth.Kerberos.Config.ValueFromSecret.Name,
-				typedTrg.Spec.Auth.Kerberos.Config.ValueFromSecret.Key,
-			)
-			secretVolumes = append(secretVolumes, configVol)
-			secretVolMounts = append(secretVolMounts, configVolMount)
-		}
+	if typedTrg.Spec.Auth != nil {
+		if typedTrg.Spec.Auth.Kerberos != nil {
+			if typedTrg.Spec.Auth.Kerberos.Config != nil {
+				configVol, configVolMount := secretVolumeAndMountAtPath(
+					"krb5-config",
+					krb5ConfPath,
+					typedTrg.Spec.Auth.Kerberos.Config.ValueFromSecret.Name,
+					typedTrg.Spec.Auth.Kerberos.Config.ValueFromSecret.Key,
+				)
+				secretVolumes = append(secretVolumes, configVol)
+				secretVolMounts = append(secretVolMounts, configVolMount)
+			}
 
-		if typedTrg.Spec.Auth.Kerberos.Keytab != nil {
-			keytabVol, keytabVolMount := secretVolumeAndMountAtPath(
-				"krb5-keytab",
-				krb5KeytabPath,
-				typedTrg.Spec.Auth.Kerberos.Keytab.ValueFromSecret.Name,
-				typedTrg.Spec.Auth.Kerberos.Keytab.ValueFromSecret.Key,
-			)
-			secretVolumes = append(secretVolumes, keytabVol)
-			secretVolMounts = append(secretVolMounts, keytabVolMount)
+			if typedTrg.Spec.Auth.Kerberos.Keytab != nil {
+				keytabVol, keytabVolMount := secretVolumeAndMountAtPath(
+					"krb5-keytab",
+					krb5KeytabPath,
+					typedTrg.Spec.Auth.Kerberos.Keytab.ValueFromSecret.Name,
+					typedTrg.Spec.Auth.Kerberos.Keytab.ValueFromSecret.Key,
+				)
+				secretVolumes = append(secretVolumes, keytabVol)
+				secretVolMounts = append(secretVolMounts, keytabVolMount)
+			}
 		}
 	}
 
@@ -126,106 +128,112 @@ func MakeAppEnv(o *v1alpha1.KafkaTarget) []corev1.EnvVar {
 			Name:  "DISCARD_CE_CONTEXT",
 			Value: strconv.FormatBool(o.Spec.DiscardCEContext),
 		},
-		{
-			Name:  envSaslEnable,
-			Value: strconv.FormatBool(o.Spec.Auth.SASLEnable),
-		},
 	}
 
-	if o.Spec.Auth.TLSEnable != nil {
-		env = append(env, corev1.EnvVar{
-			Name:  envTLSEnable,
-			Value: strconv.FormatBool(*o.Spec.Auth.TLSEnable),
-		})
-	}
+	if o.Spec.Auth != nil {
 
-	if o.Spec.Auth.SecurityMechanisms != nil {
-		env = append(env, corev1.EnvVar{
-			Name:  envSecurityMechanisms,
-			Value: *o.Spec.Auth.SecurityMechanisms,
-		})
-	}
+		if o.Spec.Auth.SASLEnable {
+			env = append(env, corev1.EnvVar{
+				Name:  envSaslEnable,
+				Value: strconv.FormatBool(o.Spec.Auth.SASLEnable),
+			})
+		}
 
-	if o.Spec.Auth.Username != nil {
-		env = append(env, corev1.EnvVar{
-			Name:  envUsername,
-			Value: *o.Spec.Auth.Username,
-		})
-	}
+		if o.Spec.Auth.TLSEnable != nil {
+			env = append(env, corev1.EnvVar{
+				Name:  envTLSEnable,
+				Value: strconv.FormatBool(*o.Spec.Auth.TLSEnable),
+			})
+		}
 
-	if o.Spec.Auth.Password != nil {
-		env = common.MaybeAppendValueFromEnvVar(
-			env, envPassword, *o.Spec.Auth.Password,
-		)
-	}
+		if o.Spec.Auth.SecurityMechanisms != nil {
+			env = append(env, corev1.EnvVar{
+				Name:  envSecurityMechanisms,
+				Value: *o.Spec.Auth.SecurityMechanisms,
+			})
+		}
 
-	if o.Spec.Auth.TLS != nil {
-		if o.Spec.Auth.TLS.CA != nil {
+		if o.Spec.Auth.Username != nil {
+			env = append(env, corev1.EnvVar{
+				Name:  envUsername,
+				Value: *o.Spec.Auth.Username,
+			})
+		}
+
+		if o.Spec.Auth.Password != nil {
 			env = common.MaybeAppendValueFromEnvVar(
-				env, envCA, *o.Spec.Auth.TLS.CA,
+				env, envPassword, *o.Spec.Auth.Password,
 			)
 		}
 
-		if o.Spec.Auth.TLS.ClientCert != nil {
-			env = common.MaybeAppendValueFromEnvVar(
-				env, envClientCert, *o.Spec.Auth.TLS.ClientCert,
-			)
+		if o.Spec.Auth.TLS != nil {
+			if o.Spec.Auth.TLS.CA != nil {
+				env = common.MaybeAppendValueFromEnvVar(
+					env, envCA, *o.Spec.Auth.TLS.CA,
+				)
+			}
+
+			if o.Spec.Auth.TLS.ClientCert != nil {
+				env = common.MaybeAppendValueFromEnvVar(
+					env, envClientCert, *o.Spec.Auth.TLS.ClientCert,
+				)
+			}
+
+			if o.Spec.Auth.TLS.ClientKey != nil {
+				env = common.MaybeAppendValueFromEnvVar(
+					env, envClientKey, *o.Spec.Auth.TLS.ClientKey,
+				)
+			}
+
+			if o.Spec.Auth.TLS.SkipVerify != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envSkipVerify,
+					Value: strconv.FormatBool(*o.Spec.Auth.TLS.SkipVerify),
+				})
+			}
 		}
 
-		if o.Spec.Auth.TLS.ClientKey != nil {
-			env = common.MaybeAppendValueFromEnvVar(
-				env, envClientKey, *o.Spec.Auth.TLS.ClientKey,
-			)
-		}
+		if o.Spec.Auth.Kerberos != nil {
+			if o.Spec.Auth.Kerberos.Config != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envKerberosConfigPath,
+					Value: krb5ConfPath,
+				})
+			}
 
-		if o.Spec.Auth.TLS.SkipVerify != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envSkipVerify,
-				Value: strconv.FormatBool(*o.Spec.Auth.TLS.SkipVerify),
-			})
-		}
-	}
+			if o.Spec.Auth.Kerberos.Keytab != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envKerberosKeytabPath,
+					Value: krb5KeytabPath,
+				})
+			}
 
-	if o.Spec.Auth.Kerberos != nil {
-		if o.Spec.Auth.Kerberos.Config != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envKerberosConfigPath,
-				Value: krb5ConfPath,
-			})
-		}
+			if o.Spec.Auth.Kerberos.ServiceName != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envKerberosServiceName,
+					Value: *o.Spec.Auth.Kerberos.ServiceName,
+				})
+			}
 
-		if o.Spec.Auth.Kerberos.Keytab != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envKerberosKeytabPath,
-				Value: krb5KeytabPath,
-			})
-		}
+			if o.Spec.Auth.Kerberos.Realm != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envKerberosRealm,
+					Value: *o.Spec.Auth.Kerberos.Realm,
+				})
+			}
 
-		if o.Spec.Auth.Kerberos.ServiceName != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envKerberosServiceName,
-				Value: *o.Spec.Auth.Kerberos.ServiceName,
-			})
-		}
+			if o.Spec.Auth.Kerberos.Username != nil {
+				env = append(env, corev1.EnvVar{
+					Name:  envKerberosUsername,
+					Value: *o.Spec.Auth.Kerberos.Username,
+				})
+			}
 
-		if o.Spec.Auth.Kerberos.Realm != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envKerberosRealm,
-				Value: *o.Spec.Auth.Kerberos.Realm,
-			})
-		}
-
-		if o.Spec.Auth.Kerberos.Username != nil {
-			env = append(env, corev1.EnvVar{
-				Name:  envKerberosUsername,
-				Value: *o.Spec.Auth.Kerberos.Username,
-			})
-		}
-
-		if o.Spec.Auth.Kerberos.Password != nil {
-			env = common.MaybeAppendValueFromEnvVar(
-				env, envKerberosPassword, *o.Spec.Auth.Kerberos.Password,
-			)
+			if o.Spec.Auth.Kerberos.Password != nil {
+				env = common.MaybeAppendValueFromEnvVar(
+					env, envKerberosPassword, *o.Spec.Auth.Kerberos.Password,
+				)
+			}
 		}
 	}
 

